Sources/SwiftBSON/BSONDocument.swift

```diff
@@ -356,12 +357,41 @@ public struct BSONDocument {
         /// Appends element to underlying BSON bytes, returns the size of the element appended: type + key + value
         @discardableResult internal mutating func append(key: String, value: BSON) -> Int {
             let writer = self.buffer.writerIndex
-            self.buffer.writeInteger(value.bsonValue.bsonType.rawValue, as: UInt8.self)
-            self.buffer.writeCString(key)
+            self.appendElementHeader(key: key, bsonType: value.bsonValue.bsonType)
             value.bsonValue.write(to: &self.buffer)
             return self.buffer.writerIndex - writer
         }
 
+        /// Append the header (key and BSONType) for a given element.
+        @discardableResult internal mutating func appendElementHeader(key: String, bsonType: BSONType) -> Int {
+            let writer = self.buffer.writerIndex
+            self.buffer.writeInteger(bsonType.rawValue, as: UInt8.self)
+            self.buffer.writeCString(key)
+            return self.buffer.writerIndex - writer
+        }
+
+        /// Build a document at the current position in the storage via the provided closure which returns
+        /// how many bytes it wrote.
+        ///
+        /// This may be used to build up a fresh document or a subdocument.
+        internal mutating func buildDocument(_ appendFunc: (inout Self) throws -> Int) rethrows -> Int {
+            var totalBytes = 0
+
+            // write length of document
+            let lengthIndex = self.buffer.writerIndex
+            totalBytes += self.buffer.writeInteger(0, endianness: .little, as: Int32.self)
+
+            // write contents
+            totalBytes += try appendFunc(&self)
+
+            // write null byte
+            totalBytes += self.buffer.writeInteger(0, as: UInt8.self)
+
+            self.buffer.setInteger(Int32(totalBytes), at: lengthIndex, endianness: .little, as: Int32.self)
```

Comment:
do we need to update the overall encoded length, too? i.e. the first 4 bytes of the storage